### PR TITLE
Autocurry feature

### PR DESCRIPTION
**Pull request by Eliah-Lakhin**
Automatic curry-ing of all undeclared variables inside anonymous functions to the function's external context by implicitly enriching function's list of parameters. This feature is optional and can be opted-out through the new `no_closures` feature.

This PR is a first part of implementation of true mutable closures.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -32,6 +32,7 @@ only_i64 = []       # set INT=i64 (default) and disable support for all other in
 no_index = []       # no arrays and indexing
 no_object = []      # no custom objects
 no_function = []    # no script-defined functions
+no_closures = []    # no automatic read/write binding of anonymous function's local variables to it's external context
```
Comment (schungx) [inline Cargo.toml:35]: Would `no_capture` be more clear about what it does?

File: src/parser.rs [hunk 2]
```diff
@@ -399,56 +402,107 @@ pub enum ReturnType {
     Exception,
 }
 
-#[derive(Clone)]
```
Comment (schungx) [inline src/parser.rs:402]: Why remove derive `Clone`?

File: src/parser.rs [hunk 3]
```diff
@@ -399,56 +402,107 @@ pub enum ReturnType {
...
 struct ParseState<'e> {
-    /// Reference to the scripting `Engine`.
```
Comment (schungx) [inline src/parser.rs:404]: I suggest leave them at three slashes... this style is consistent with the rest of the code.  If we eventually make this type public, then it automatically generates documentation.

**Comment by schungx** [comment]
I suggested a number of stylistic changes.  Otherwise this is ready to merge.

**Comment by Eliah-Lakhin** [comment]
@schungx Done. All issues fixed.

**Comment by schungx** [comment]
Great.  Thanks.  I'll write up the documentation in the book (or would you like to do that)?

**Comment by schungx** [comment]
I did some minor style refactors, then put in the correct position for variables not found.

**Comment by Eliah-Lakhin** [comment]
@schungx I suppose the end-user can bypass this issue by introducing a local variable explicitly and without init value like this:

```
let f = |y| { let z; eval("let z = 0"); x + y + z };
```

It will put some additional overhead, but not too expensive I think. Anyway, the `no_capture` feature was exactly for such situations: if the user don't need or don't want such syntax-suger he can turn it off. Any syntax-sugar has drawbacks.

Furthermore we can introduce a new keyword such as `local z;` that will explicitly say that the variable is supposed to be somehow present in the anon function scope's so it should be bound to the external context, but the `Engine` will not interpret it anyhow, it will be for the parsing stage only.

**Comment by schungx** [comment]
Well, I guess let's leave it at this.  I don't think it is too much of a problem, because the only way for a variable access within a function not to throw an error is through an `eval` that defines that variable.

So throwing an error earlier might not matter much -- unless the user uses `eval`, he/she gets an error anyway when running the function.

And if the user uses `eval`, then I guess he/she just needs to be careful!

**Comment by schungx** [comment]
@Eliah-Lakhin   I find that your addition of `externals` open up yet other possibilities.  For example, it is now possible to make special function calls such that functions can refer to variables in the execution scope.

I'm planning to add something like this:

```rust
fn foo(x) { y += x; }

let y = 41;

foo(1);      // error - 'y' not found

foo!(1);    // bang (similar to Rust's macros) means execute in the current context
               // 'y' is in foo's externals set, so it is copied into the scope for this call
               // and then updated back to the scope after the call

y == 42;
```

**Comment by schungx** [comment]
The latest `master` has this feature.  You can do `func!(...)` to call the function within the current scope.

I ended up having to clone the entire `Scope` object in order to please the borrow checker.  Function resolution doesn't happen early enough (since it depends on the types of arguments) for us to cherry-pick the variables to clone.  By the time the argument types are obtained, the `Scope` is already locked by a mutable reference to the first argument and it is no longer possible to get values out of it.

**Pull request merged**